File: sdks/node/packages/sdk/src/adapter.ts
import type { FireweaveError } from './errors.js';
import type { CanonicalContext, DecisionReason, Exposure, FlagValueType, JsonValue, Signal } from './types.js';

/** What is being registered — see spec/remote-register-target.schema.json. */
export type TargetKind = 'user' | 'device';

export interface RegisterTargetOptions {
  /** Defaults to 'user'. */
  kind?: TargetKind;
  /** Durable targeting facts: plan, beta membership, region, device model. */
  properties?: Record<string, JsonValue>;
  /** Client-declared environment (production, staging, …). */
  environment?: string;
  signal?: AbortSignal;
}

export interface RegisterTargetResult {
  /**
   * `false` means the target was NOT registered — rules that depend on its
   * properties will not match until a later attempt succeeds. Callers in a
   * login path normally ignore this; a careful caller logs it, because a
   * silently unregistered target is exactly how targeting rules end up
   * matching nobody.
   */
  readonly ok: boolean;
  readonly error?: FireweaveError;
}

/** Result of a backend flag resolution (success path; faults throw FireweaveError). */
export interface AdapterResolution {
  found: boolean;
  /** Flag enabled state; false ⇒ DISABLED reason with the flag's stored value. */
  enabled?: boolean;
  value?: JsonValue;
  variant?: string;
  /** Declared flag type when known (in-memory fixtures); adapters may omit. */
  flagType?: FlagValueType | 'integer' | 'float';
  /**
   * Adapter-suggested reason override (e.g. SPLIT, STALE) on a `found: true`
   * resolution. On a `found: false` miss, `reason: 'DEFAULT'` instead signals
   * the runtime to return the caller's default with reason `DEFAULT` — not an
   * error — per local mode's unknown-key row (spec/modes.md; see
   * FireweaveLocalAdapter.resolve).
   */
  reason?: DecisionReason;
  /** Vendor reason code (e.g. "condition_match"). */
  reasonCode?: string;
  /** Vendor condition index when the vendor reports one. */
  conditionIndex?: number;
  /** Flag definition version. */
  version?: number;
  /** Vendor-side numeric flag id. */
  vendorFlagId?: number;
  /** Optional flag payload (object or pre-serialized JSON string). */
  payload?: JsonValue;
  /** True when the /flags response reported quota limiting. */
  quotaLimited?: boolean;
  /** True when served from a stale/last-good cache. */
  fromCache?: boolean;
}

export interface ResolveOptions {
  signal?: AbortSignal;
}

export interface AdapterRuntimeFeatures {
  remoteEvaluation?: boolean;
  localEvaluation?: boolean;
  localOnly?: boolean;
  exposureEmission?: boolean;
  sideEffectFreeReads?: boolean;
  groupAnalytics?: boolean;
}

/**
 * Backend adapter boundary (docs/architecture.md §layers). Adapters translate
 * canonical requests to vendor protocols; they never see OpenFeature types.
 */
export interface BackendAdapter {
  /** Third-party adapters use 'other'; the vendor is fw-server's concern. */
  readonly name: 'inmemory' | 'fireweave' | 'other';
  /** Bring the backend to a usable state. Reject with FireweaveError on failure. */
  initialize(signal?: AbortSignal): Promise<void>;
  /** Resolve one flag. Throws FireweaveError for transport/auth/parse faults. */
  resolve(flagKey: string, context: CanonicalContext, options?: ResolveOptions): Promise<AdapterResolution>;
  /**
   * Register a user or device so rules can target its durable properties
   * (optional capability). Resolves with `ok: false` rather than throwing —
   * registration sits in login paths and must not break sign-in.
   */
  registerTarget?(
    targetingKey: string,
    options?: RegisterTargetOptions,
  ): Promise<RegisterTargetResult>;
  /** Record an exposure event (optional capability). */
  recordExposure?(exposure: Exposure): void;
  /** Deliver a telemetry signal to the backend sink (optional capability, ruling 17). */
  recordSignal?(signal: Signal): void;
  /** Flush buffered telemetry/exposures. */
  flush?(): Promise<void>;
  /** Release resources; must be idempotent. */
  shutdown(): Promise<void>;
  /** Runtime feature capabilities contributed by this adapter. */
  features(): AdapterRuntimeFeatures;
}

File: sdks/go/domain/adapter.go
package domain

import "context"

// ResolveRequest carries one flag resolution through the runtime to a
// BackendAdapter. Context has already been merged and validated and is a
// private copy the adapter may read freely.
type ResolveRequest struct {
	FlagKey      string
	Type         FlagType
	DefaultValue any
	Context      EvaluationContext
}

// BackendAdapter is the vendor seam. Implementations:
// infrastructure/adapters/remote (production fw-server path),
// infrastructure/adapters/local (offline development),
// infrastructure/adapters/inmemory (deterministic fixture adapter for
// tests).
//
// This port lives in domain rather than application deliberately: Go
// resolves import dependencies per PACKAGE (unlike node/python/java, whose
// module systems resolve per FILE), and application's composition root
// (application/mode.go) must import the concrete adapter packages to wire
// Init's adapter selection. If the port types lived in application instead,
// every adapter package would need to import application for them — and
// application (via mode.go) importing those same adapter packages back
// would be a package-level import cycle, which Go rejects at compile time.
// Placing the port in domain (imported by everyone, importing nothing
// outside domain) is the Go-idiomatic resolution: this is exactly the kind
// of import-cycle-driven layout adjustment node/java did not need to make.
//
// Implementations must be safe for concurrent use: Resolve may be called
// from many goroutines, potentially concurrently with Close. Every blocking
// method accepts a context.Context and must honor cancellation and
// deadlines.
type BackendAdapter interface {
	// Initialize prepares the adapter (client construction, initial polls).
	// A returned *Error with KindConfiguration is treated as fatal.
	Initialize(ctx context.Context) error
	// Resolve evaluates one flag. Failures are reported inside the
	// Decision (Reason ERROR + typed Error) or, for an adapter whose
	// "unknown key" row is default/DEFAULT rather than an error
	// (spec/modes.md "Behaviour per mode"), as a plain Decision carrying
	// the caller's default with reason DEFAULT — Resolve never panics.
	Resolve(ctx context.Context, req ResolveRequest) Decision
	// Close releases resources. It must be idempotent and must return once
	// ctx is done even if the underlying vendor client would wait longer.
	Close(ctx context.Context) error
}

// RegisterTargetOptions carries the optional fields for
// POST /v1/targets/register (spec/remote-register-target.schema.json).
type RegisterTargetOptions struct {
	// Kind defaults to "user" (TargetKindUser) when empty.
	Kind TargetKind
	// Properties are durable targeting facts: plan, beta membership,
	// region, device model.
	Properties map[string]any
	// Environment is the client-declared environment (production, staging, …).
	Environment string
}

// RegisterTargetResult reports the outcome of registration. OK false means
// the target was NOT registered — rules that depend on its properties will
// not match until a later attempt succeeds. Callers in a login path
// normally ignore this; a careful caller logs it, because a silently
// unregistered target is exactly how targeting rules end up matching
// nobody.
type RegisterTargetResult struct {
	OK    bool
	Error *Error
}

// TargetRegistrar is optionally implemented by adapters that can register a
// target (infrastructure/adapters/local records in-process and traces;
// infrastructure/adapters/remote POSTs to fw-server). Runtime discovers it
// via type assertion; an adapter that does not implement it (e.g. the
// fixture-only inmemory adapter) degrades RegisterTarget with
// UnsupportedCapability — never a panic, since registration sits in
// sign-in paths where a targeting concern must not break authentication.
type TargetRegistrar interface {
	RegisterTarget(ctx context.Context, targetingKey string, opts RegisterTargetOptions) RegisterTargetResult
}
